%\VignetteIndexEntry{R classes and methods for Illumina bead-based data}
%\VignetteDepends{}
%\VignetteKeywords{beadarray expression analysis}
%\VignettePackage{beadarray}
\documentclass{article}
\usepackage{hyperref}

\textwidth=6.2in
\textheight=8.5in
%\parskip=.3cm
\oddsidemargin=.1in
\evensidemargin=.1in
\headheight=-.3in

\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textsf{#1}}}
\newcommand{\Rmethod}[1]{{\texttt{#1}}}
\newcommand{\Rfunarg}[1]{{\texttt{#1}}}
\newcommand{\Rclass}[1]{{\textit{#1}}}

\begin{document}
\title{\Rpackage{beadarray}: R classes and methods for Illumina bead-based data}
\author{Mark Dunning, Mike Smith, Johnathan Cairns, Andy Lynch and Matt Ritchie}
\maketitle



\section*{Introduction}

\Rpackage{beadarray} is a Bioconductor package for the analysis of Illumina data. 
The main feature of the package is its ability to process raw data (text files and TIFFs) from 
BeadScan and store this information in convenient R classes.
Access to this \emph{bead-level data} offers users the choice between different image processing 
and background correction methods and allows for detailed quality assessments to be carried out for 
each array. The summarised intensities (\emph{bead-summary data}) output by Illumina's BeadStudio software can 
also be imported using the \Rpackage{beadarray} package.

The documentation for this package is split into separate user guides: one for bead-level data and 
a second for bead-summary data.
These documents may be accessed using the following commands:

<<Help, eval=FALSE>>=
library("beadarray")
beadarrayUsersGuide(topic="beadlevel")
beadarrayUsersGuide(topic="beadsummary")
@

\section*{Changes since Bioconductor 2.6 (beadarray 1.16)}

There has been a major overhaul of \Rpackage{beadarray} since Bioconductor 2.6 with the intention of creating a more flexible package. Users now have more flexibility in the way quality assessment is performed at the bead-level and are allowed to modify the raw bead-level object. Summarisation is also more flexible and retrieval of annotation data can be performed from within the package. See the package vignettes for details and the package news for updates.

<<NEWS, eval=FALSE, echo=TRUE>>=
news(package="beadarray")
@

\section*{Companion packages to beadarray}

In order to reduce the download size of \Rpackage{beadarray}, we have moved the example dataset to a Bioconductor data package (\Rpackage{beadarrayExampleData}) that can be downloaded in the usual way. An expanded dataset and more detailed commentary on the processing of BeadArray data is provided by the \Rpackage{BeadArrayUseCases} package. Finally, the \Rpackage{beadarray} package is also able to interact with various Illumina annotation packages in Bioconductor if they are installed. 

<<CompanionPkg, eval=FALSE>>=

source("http://www.bioconductor.org/biocLite.R")

biocLite(c("beadarrayExampleData", "beadarrayExampleData"))

@


\section*{Citing \Rpackage{beadarray}}

\begin{flushleft}
If you use \Rpackage{beadarray} for the analysis or pre-processing of
Illumina data please cite:\newline

Dunning MJ, Smith ML, Ritchie ME, Tavar\'e S. \emph{beadarray: R classes and
methods for Illumina bead-based data}. Bioinformatics, 2007 Aug 15; 23(16):2183-4.\newline

For recommendations on the low-level analysis of bead-level data, or if you make use of the spike-in data set available from\newline

{\tt http://www.compbio.group.cam.ac.uk/Resources/spike/}\newline

please cite:\newline

Dunning MJ, Barbosa-Morais NL, Lynch AG, Tavar\'e S, Ritchie ME. \emph{Statistical issues in the analysis of Illumina data}. 
BMC Bioinformatics. 2008 Feb 6;9(1):85.\newline

For a validation of the VST method (implemented in the \Rpackage{lumi} package) on the spike-in data set and a comparison 
of the results obtained from a full bead-level analysis, see\newline

Dunning MJ, Ritchie ME, Barbosa-Morais NL, Tavar\'e S, Lynch AG. \emph{Spike-in validation of an Illumina-specific 
variance-stabilizing transformation}. BMC Research Notes, 2008, 1:18.\newline

If you use BASH to detect spatial artefacts, please cite\newline

Cairns JM, Dunning MJ, Ritchie ME, Russell R, Lynch AG. \emph{BASH: A tool for managing BeadArray spatial artefacts}. Bioinformatics, 2008, 15:24(24).\newline

If you use the probe reannotation for Illumina expression arrays (used in the current Bioconductor annotation packages named \Rpackage{illuminaXXXXX.db}) 
please cite\newline

Barbosa-Morais NL, Dunning MJ, Samarajiwa SA, Darot JF, Ritchie ME, Lynch AG, Tavar\'e S. \emph{A re-annotation pipeline for Illumina BeadArrays: improving the interpretation of gene expression data.}. Nucleic Acids Research 2010 Jan;38(3)e17.Epub 2009 Nov 18.\newline

Supplementary tables and supporting information for this annotation are available at\newline

{\tt http://www.compbio.group.cam.ac.uk/Resources/Annotation}.

\end{flushleft}

\section{Asking for help on \Rpackage{beadarray}}

Wherever possible, please send all queries about \Rpackage{beadarray} to the
Bioconductor mailing list at {\tt bioconductor@stat.math.ethz.ch}. This will
help maintain a searchable archive of questions and responses.
When posting to the list, please include the commands you used along with the
version of \Rpackage{beadarray} and {\tt R} you are working with.
Version information can be obtained by running the following command:

<<Mailing list, eval=FALSE>>=
sessionInfo()
@

\end{document}
